File: src/engine/math.ts
import type { Agent } from './types';

export type PreferenceParams = {
  readonly alpha: number;
  readonly beta: number;
};

/**
 * Derives the scaling parameters for the Cobb-Douglas utility function.
 * alpha: Weight for food (needs to scale based on Hunger)
 * beta: Weight for gold (scales based on base Greed)
 */
export const derivePreferenceParams = (agent: Agent): PreferenceParams => {
  // E.g. A hyper-hungry agent sees a massive spike in alpha, overriding beta.
  const alpha = Math.pow(10, agent.currHunger); // Hunger.Hi (2) -> 100, Lo (1) -> 10, None (0) -> 1
  const beta = Math.pow(10, agent.baseGreed) * 0.5; // Scaled so Hunger prevents starvation
  return { alpha, beta };
};

/**
 * Calculates current Marginal Rate of Substitution (MRS) given an agent's state.
 * Represents how many units of Gold the agent values 1 unit of Food at.
 * MRS = - (MU_F / MU_G) = (alpha / beta) * ((G + 1) / (F + 1))
 */
export const calculateMRS = (agent: Agent): number => {
  const { alpha, beta } = derivePreferenceParams(agent);
  const { foodInventory: f, goldInventory: g } = agent;
  
  return (alpha / beta) * ((g + 1) / (f + 1));
};

/**
 * Calculates the total utility of a given consumption bundle
 * U(F, G) = (F + 1)^alpha * (G + 1)^beta
 */
export const calculateUtility = (food: number, gold: number, params: PreferenceParams): number => {
  return Math.pow(food + 1, params.alpha) * Math.pow(gold + 1, params.beta);
};

/**
 * Generates a set of (Food, Gold) points that provide the exact same utility 
 * as the agent's current inventory. These are used to plot the indifference curve.
 */
export const generateIndifferenceCurve = (agent: Agent, maxFood: number = 20): Array<{x: number, y: number}> => {
  const params = derivePreferenceParams(agent);
  const currentUtility = calculateUtility(agent.foodInventory, agent.goldInventory, params);
  const curve = [];

  for (let f = 0; f <= maxFood; f++) {
    // Math derivation:
    // U = (F + 1)^alpha * (G + 1)^beta
    // (G + 1)^beta = U / (F + 1)^alpha
    // G + 1 = (U / (F + 1)^alpha)^(1 / beta)
    // G = (U / (F + 1)^alpha)^(1 / beta) - 1
    
    // We max(0, g) to prevent negative gold points in the plot 
    const base = currentUtility / Math.pow(f + 1, params.alpha);
    const g = Math.pow(base, 1 / params.beta) - 1;
    curve.push({ x: f, y: Math.max(0, g) });
  }

  return curve;
};

File: econ-sim-elixir/lib/econ_sim/math.ex
defmodule EconSim.Math do
  @moduledoc """
  Microeconomic math functions: Cobb-Douglas utility, MRS, marginal utilities,
  and indifference curve generation. Direct port of src/engine/math.ts.
  """

  alias EconSim.Types.Agent

  @type preference_params :: %{alpha: float(), beta: float()}

  @doc """
  Derives the Cobb-Douglas scaling parameters from an agent's hunger and greed.
  alpha: weight for food (scales exponentially with hunger)
  beta: weight for gold (scales exponentially with greed)
  Normalized so alpha + beta = 1.
  """
  @spec derive_preference_params(Agent.t()) :: preference_params()
  def derive_preference_params(%Agent{} = agent) do
    raw_alpha = :math.pow(2, agent.curr_hunger)
    raw_beta = :math.pow(2, agent.base_greed) * 0.5
    sum = raw_alpha + raw_beta
    %{alpha: raw_alpha / sum, beta: raw_beta / sum}
  end

  @doc """
  Calculates Marginal Rate of Substitution.
  MRS = (alpha / beta) * ((G + 1) / (F + 1))
  Represents how many gold units the agent values 1 food unit at.
  """
  @spec calculate_mrs(Agent.t()) :: float()
  def calculate_mrs(%Agent{} = agent) do
    %{alpha: alpha, beta: beta} = derive_preference_params(agent)
    (alpha / beta) * ((agent.gold_inventory + 1) / (agent.food_inventory + 1))
  end

  @doc """
  Marginal utility of food.
  MU_F = alpha * (F + 1)^(alpha - 1) * (G + 1)^beta
  """
  @spec calculate_mu_food(Agent.t()) :: float()
  def calculate_mu_food(%Agent{} = agent) do
    %{alpha: alpha, beta: beta} = derive_preference_params(agent)
    f = agent.food_inventory
    g = agent.gold_inventory
    alpha * :math.pow(f + 1, alpha - 1) * :math.pow(g + 1, beta)
  end

  @doc """
  Marginal utility of gold.
  MU_G = beta * (F + 1)^alpha * (G + 1)^(beta - 1)
  """
  @spec calculate_mu_gold(Agent.t()) :: float()
  def calculate_mu_gold(%Agent{} = agent) do
    %{alpha: alpha, beta: beta} = derive_preference_params(agent)
    f = agent.food_inventory
    g = agent.gold_inventory
    beta * :math.pow(f + 1, alpha) * :math.pow(g + 1, beta - 1)
  end

  @doc """
  Calculates total utility of a consumption bundle.
  U(F, G) = (F + 1)^alpha * (G + 1)^beta
  """
  @spec calculate_utility(float(), float(), preference_params()) :: float()
  def calculate_utility(food, gold, %{alpha: alpha, beta: beta}) do
    :math.pow(food + 1, alpha) * :math.pow(gold + 1, beta)
  end

  @doc """
  Generates indifference curve points for the agent's current utility level.
  Returns a list of %{x: food, y: gold} points.
  """
  @spec generate_indifference_curve(Agent.t(), integer()) :: [%{x: float(), y: float()}]
  def generate_indifference_curve(%Agent{} = agent, max_food \\ 20) do
    params = derive_preference_params(agent)
    current_utility = calculate_utility(agent.food_inventory, agent.gold_inventory, params)

    Enum.map(0..max_food, fn f ->
      # G = (U / (F + 1)^alpha)^(1 / beta) - 1
      base = current_utility / :math.pow(f + 1, params.alpha)
      g = :math.pow(base, 1 / params.beta) - 1
      %{x: f, y: max(0.0, g)}
    end)
  end
end
